### PR TITLE
build.rs: remove target darknet source files right before copying them

### PR DESCRIPTION
**Pull request by PabloMansanet**
This solves the `OS error 13` permissions problem, by allowing the build script to clobber read-only files when copying libraries across.

In particular, git `pack` files are created with `555` permissions, which means they weren't being properly overwritten by the build script, which resulted in the following build failures:
```
error: failed to run custom build command for `darknet-sys v0.3.1 ([https://github.com/tonarino/darknet-sys-rust?branch=openmp-configurable#ed9b0b4a)`](https://github.com/tonarino/darknet-sys-rust?branch=openmp-configurable#ed9b0b4a)%60)

Caused by:
  process didn't exit successfully: `/home/strohel/work/portal/target/debug/build/darknet-sys-13f789a883e167c3/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=DARKNET_SRC
  cargo:rerun-if-env-changed=DARKNET_INCLUDE_PATH
  cargo:rerun-if-env-changed=CUDA_PATH
  cargo:rerun-if-env-changed=CUDA_ARCHITECTURES
  cargo:rerun-if-env-changed=DARKNET_ENABLE_OPENMP
  cargo:rerun-if-env-changed=/home/strohel/work/portal/target/debug/build/darknet-sys-7ad42202ae742168/out/bindings.rs
  process: "/home/strohel/.cargo/git/checkouts/darknet-sys-rust-cf3c15bddd28ea3e/ed9b0b4/darknet"
    copy: "/home/strohel/.cargo/git/checkouts/darknet-sys-rust-cf3c15bddd28ea3e/ed9b0b4/darknet/.gitignore" -> "/home/strohel/work/portal/target/debug/build/darknet-sys-7ad42202ae742168/out/darknet/.gitignore"
  (...)
    copy: "/home/strohel/.cargo/git/checkouts/darknet-sys-rust-cf3c15bddd28ea3e/ed9b0b4/darknet/.git/refs/heads/master" -> "/home/strohel/work/portal/target/debug/build/darknet-sys-7ad42202ae742168/out/darknet/.git/refs/heads/master"
  process: "/home/strohel/.cargo/git/checkouts/darknet-sys-rust-cf3c15bddd28ea3e/ed9b0b4/darknet/.git/objects"
  process: "/home/strohel/.cargo/git/checkouts/darknet-sys-rust-cf3c15bddd28ea3e/ed9b0b4/darknet/.git/objects/pack"
    copy: "/home/strohel/.cargo/git/checkouts/darknet-sys-rust-cf3c15bddd28ea3e/ed9b0b4/darknet/.git/objects/pack/pack-83da3357e7e355ccb5a09569a7336054d38eba2d.idx" -> "/home/strohel/work/portal/target/debug/build/darknet-sys-7ad42202ae742168/out/darknet/.git/objects/pack/pack-83da3357e7e355ccb5a09569a7336054d38eba2d.idx"

  --- stderr
  Error: Permission denied (os error 13)
warning: build failed, waiting for other jobs to finish...
error: build failed
```

---
This is all independent from any error readability improvements, that will be part of a separate PR.

### CODE REVIEW

File: build.rs
Comment (bschwind) [inline build.rs]: Minor nitpick, but I'm usually always against `bool` arguments, as you always have to go to the function definition to see what `true` or `false` mean.

Comment (PabloMansanet) [reply build.rs]: Good point, agreed.

I'm not sure how to solve this in a general sense on a single-file script like this though. With some encapsulation, I typically write an "impl" style function in an inner module, and wrap it with appropriately variants covering every `bool` possibility. Is it worth it doing that on a `build.rs` script you think? Alternatively I can have an enum for the argument, which may be simpler here.

Comment (bschwind) [reply build.rs]: Does the `copy` function actually need to be marked `pub`? If not, then just defining a simple `enum` in `build.rs` seems pretty straightforward unless I'm missing something.

Comment (PabloMansanet) [reply build.rs]: I'll go with the `enum` choice then :)

Comment (bschwind) [reply build.rs]: Thanks!

**Review by bschwind — APPROVED** [review]
LGTM, I'll defer to @strohel for further review

File: build.rs
Comment (strohel) [inline build.rs]:
```suggestion
                        // Some `git` files are created with write protection, so replacing them
                        // directly can fail with a permissions error. Remove the destination file
                        // first. Ignore any errors, only the fs::copy() call is critical.
                        fs::remove_file(&dest_path).ok();
```

I've realized we only call `copy()` from one place, always with `OverwriteBehaviour::ClobberReadOnlyFiles`, so I don't think that pulls the weight of introducing an enum and a paremeter -> I propose to simply hardcode the behavior. This is a single-purpose build script, so this can be arbitrarily specific.

If we fear this behavior is surprising, we can document that in `copy()`'s docstring.

**Review by strohel — CHANGES_REQUESTED** [review]
Thanks for digging out (and fixing) the cause!

In code I propose not to introduce the parameter at all.

BTW, once we're happy with this, we should submit this fix upstream (instead of merging). The author has already merged 2 our PRs in the past.

Until both PRs are merged upstream, we can create an integration branch in this fork to combine the 2 branches and reference that from portal.

File: build.rs
Comment (bschwind) [reply build.rs]: Agreed, when I saw the bool parameter I didn't realize it was only called once. The `pub` marking also threw me off.

Comment (PabloMansanet) [reply build.rs]: Yeah, I was thrown off by the pub too 😅 I agree about making the behaviour inconditional.

Comment (strohel) [reply build.rs]: I went ahead and pushed this change.

**Review by strohel — APPROVED** [review]
I think this is ready for submitting upstream now.

@PabloMansanet do you want to do it, or should I?

**Comment by PabloMansanet** [comment]
Sorry, it completely slipped my mind to continue work in this PR :sweat: thanks for taking care of that! 

Please go ahead and PR it upstream :) I don't have a lot of context about darknet as a whole so if they ask me any hard questions I'm going to be unable to justify this very well!

**Comment by strohel** [comment]
Filled in as https://github.com/alianse777/darknet-sys-rust/pull/15, closing this one.

**Pull request closed without merging**
